### PR TITLE
add has words first max and fixes cors

### LINKED ISSUE #25068: Add `hasWords`, first and max to messages endpoint
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/api

### Describe the bug

In order to build the new screen that we have planned for Internationalisation, we need a filter for translations that have specific words and a first and max.

### PR DESCRIPTION
**Pull request by edewit**
Signed-off-by: Erik Jan de Wit <erikjan.dewit@gmail.com>

fixes: #25068

### CODE REVIEW

**Review by agagancarczyk — DISMISSED** [review]
LGTM!

File: services/src/main/java/org/keycloak/services/resources/ThemeResource.java [hunk 1]
```diff
@@ -107,7 +108,9 @@ public Response getResource(@PathParam("version") String version, @PathParam("th
     @Produces(MediaType.APPLICATION_JSON)
     public Response getLocalizationTexts(@PathParam("realm") String realmName, @QueryParam("theme") String theme,
                                          @PathParam("locale") String localeString, @PathParam("themeType") String themeType,
-                                         @QueryParam("source") boolean showSource) throws IOException {
+                                         @QueryParam("source") boolean showSource, @QueryParam("hasWords") List<String> hasWords,
+                                         @QueryParam("first") @DefaultValue("0") long first,
+                                         @QueryParam("max") long max) throws IOException {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/resources/ThemeResource.java:113]: Please use `Integer` instead of `long`, and account for `null` in the code.

```suggestion
                                         @QueryParam("first") Integer first,
                                         @QueryParam("max") Integer max) throws IOException {
```

File: services/src/main/java/org/keycloak/services/resources/ThemeResource.java [hunk 2]
```diff
@@ -138,8 +141,17 @@ public Response getLocalizationTexts(@PathParam("realm") String realmName, @Quer
                     new KeySource((String) e.getKey(), (String) e.getValue())).collect(toList());
         }
 
+        if (!hasWords.isEmpty()) {
+            result = result.stream().filter(keySource -> hasWords.stream()
+                    .anyMatch(w -> keySource.getValue().toUpperCase(locale).contains(w.toUpperCase(locale)))).collect(toList());
+        }
+
+        if (max != 0) {
+            result = result.stream().skip(first).limit(max).collect(toList());
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/resources/ThemeResource.java:150]: In Keycloak REST API it is common to have `first` and `max` settable independently.

This part of code converts a stream to list and back several times which increases pressure on GC due to number of created fresh objects. Instead, keep the stream only and collect at the end. That way it is also possible to have `.skip` and `.limit` applied selectively when the `first` / `max` are non-null, and similarly `hasWords` parameter.

File: services/src/main/java/org/keycloak/services/resources/ThemeResource.java [hunk 3]
```diff
@@ -107,7 +108,9 @@ public Response getResource(@PathParam("version") String version, @PathParam("th
     @Produces(MediaType.APPLICATION_JSON)
     public Response getLocalizationTexts(@PathParam("realm") String realmName, @QueryParam("theme") String theme,
                                          @PathParam("locale") String localeString, @PathParam("themeType") String themeType,
-                                         @QueryParam("source") boolean showSource) throws IOException {
+                                         @QueryParam("source") boolean showSource, @QueryParam("hasWords") List<String> hasWords,
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/resources/ThemeResource.java:111]: I'd suggest changing the `@QueryParam("hasWords")` to `@QueryParam("q")`  and changing its type to `String` similarly to other free-text search fields in REST API, e.g. [here](https://github.com/keycloak/keycloak/blob/e13d3264a2c770f390460bf0db9247e1bf77646a/services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java#L404) or [here](https://github.com/keycloak/keycloak/blob/e13d3264a2c770f390460bf0db9247e1bf77646a/services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java#L88).

Then it would be possible to process the `q` field into a `Set<String>` which would contain non-empty and non-null elements trimmed of whitespace. Having a `Set` instead of `List` would allow simplification of the condition on L 146.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
@edewit Could you please add a testcase for the updated functionality into `ThemeResourceProviderTest`? The testcase should cover both including and not including the first / max / words.

Could you link an issue to this PR?

File: services/src/main/java/org/keycloak/services/resources/ThemeResource.java (hunk 1 above)
Comment (edewit) [reply services/src/main/java/org/keycloak/services/resources/ThemeResource.java:113]: the signature of skip is `long` though

File: services/src/main/java/org/keycloak/services/resources/ThemeResource.java (hunk 3 above)
Comment (edewit) [reply services/src/main/java/org/keycloak/services/resources/ThemeResource.java:111]: These examples the searchQuery has the form `name:value`, but here it's just a number of words that have to be in the message bundle these are no attribute names.

**Review by agagancarczyk — DISMISSED** [review]

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you for the additional test case. Unfortunately, it revealed some issues that need to be addressed. Please see inline.

The pagination is not stable - the order of returned items can alter across calls. Also the test coverage needs be higher and include negative test cases, and the endpoint should fail gracefully on invalid input.

I have prepared a PR into your repository with fixes. Feel free to use it: https://github.com/edewit/keycloak/pull/5
